### DISCUSSION TITLE
access /proc causes problems on shared hosting

### DISCUSSION THREAD

**Discussion by drdingens** [akeeba/panopticon#834] [Q&A (Community Support)]
Hey
I got a complaint from my hoster, saying:
"You have software on the web space that is trying to access /proc and it is flooding quite a lot of requests, which is why the log is no longer really readable. Access to /proc is not possible in shared hosting, as this is the server's resource management and is reserved for the root user."

They also send this screenshot:
![2024_10_25_20_53_40_Mail](https://github.com/user-attachments/assets/01a37da7-faca-4c37-ad77-f35a1c2ef828)

I deactivated all sites in Panopticon and they said now everything is fine again.
Is there a way to avoid this?
Thanks
Rolf

**Comment by nikosdion** [comment]
Something is fishy with the host's reply. Please [do read what `/proc` actually is and what it does](https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/proc.html).

The connector reads exactly two things from the `/proc` virtual filesystem: 
1. `/proc/stat` to get information about the CPU usage. This takes about 1msec.
2. `/proc/cpuinfo` to get information about the CPU itself. This takes about 3msec.

The idea is that we use it to collect serer performance information which is actually useful _for you_, the site owner. Your site may be dog slow all of a sudden. What do you do? Normally, you'd have to go into the hosting control panel and inspect the system load (which only makes sense if you know how many compute cores you have), the CPU usage, and memory usage. The problem is that the hosting control panel will only report system load without the CPU information, the CPU usage reported is only the _user_ portion of the CPU used, and the memory load reported is usually the total memory usage _including the swap space_. Here's why each one of these can be misleading.

The system load is a float, e.g. 1.23. What does that mean? By itself, nothing. If you tell me this is on a single core machine, it tells me that you are are at 123% capacity, therefore processes are lagging behind, i.e. you need a bigger server. If you tell me that's on a 16 core CPU I'll tell you that you need to check the actual CPU usage stats to see what's going on.

CPU usage in Linux will be divided between user, system, iowait, and idle (there are a few more categories you can reasonably put into system). The system use is what the OS is using and you have no control over. This is the overhead you can't address at your level, but it's also CPU power you can't use. The user load is what the userspace applications like your web server, PHP, the database etc consume. This is what you can control depending on how efficiently your software is, and how your site is built. The idle load is, well, available compute power not being used. The iowait is very important, though; it's the amount of time the OS kernel is doing nothing because it's waiting for I/O. If this spikes above 5% (and especially if it goes over 10%) your server is struggling to keep up because of slow storage or slow network. You may have plenty of idle CPU time but your server will be dog slow because the OS cannot do anything useful as it's bottlenecked by its storage and/or network. This is a huge problem, and I see why the host might not want you to see.

By default, `/proc` is not limited to the root user. Its default permissions are 0777, and the permissions of the virtual files `stat` and `cpuinfo` in it are 0444, meaning it's readable by everyone. It _can_ be limited to root using mount parameters (see [this kernel commit](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/diff/Documentation/filesystems/proc.txt?id=0499680a42141d86417a8fbaa8c8db806bea1201&id2=97412950b10e64f347aec4a9b759395c2465adf6)), but in this case we just get an immediate fopen() failure without troubling the kernel. So, somehow linking that to something you are not quite giving me the context of sounds very fishy to me.

The only thing I can imagine being affected by trying to access what you're not allowed is log entries in the system log. But these are by no means as many or as verbose as the onslaught of 404s your server logs constantly by browsers looking for but not finding the favicon.ico, for example.

I actually need more context from you about WHAT the problem you have actually is. What log is no longer readable? When is that log rotated? What is its retention policy? Do they just bump on a storage limitation? How many requests do you get for accessing `/proc`? How many web requests you get in total over those sites? If we're talking about a relative log rate of 0.01% then the problem isn't really accessing `/proc`, is it?

**Comment by drdingens** [comment]
Hey
All I have so far is that the server had performance issues since May and some time outs when I monitored it with Hetrix Tools. The webhoster escalated it to the tech team and they couldn't really pin it down. We were about to move to another server, when one guy came up with the observation I quoted above.
After I turned off Panopticon, they said:
"as of today I no longer see any messages about access to /proc being attempted, so the deactivation was successful. We keep an eye on the server to see if it is of any help". That was yesterday, so it is a bit early to tell if it was of any help or not.
Will ask the webhoster team again with the info you gave me, thanks a lot!
Rolf

**Reply by nikosdion** [reply]
Do you have HetrixTools' server stats agent enabled? That script collects _the exact same information, in the exact same way (accessing `/proc`)_ as Panopticon, but far more frequently: once every minute per site, instead of once ever 15 to 60 minutes in Panopticon. I know because I've been running that on my production sites for years.

Their excuses sound extremely fishy to me. It sounds like they have not set up a journald remote server for their fleet, instead storing the entire system journal locally, and with the default retention policy(see [journald log rotation](https://wiki.archlinux.org/title/Systemd/Journal#Journal_size_limit)). This will of course cause the log to balloon to epic proportions, slowing down the server, up to the point they run out of local storage and the whole thing grinds to a screeching halt. This is a server management issue, not an issue with collecting server performance information.

Also keep in mind that GOOD hosts have no problems making `/proc` read-only for all local users. A GOOD host has nothing to hide about the actual performance of their server. They are proactive. They monitor these parameters and will take action whenever the server performance gets outside the expected operating parameters as this could signify a major problem such as a broken site, a DoS attack, a database issue, etc. The only hosts who will hide performance metrics from you are those who know they are running an overcrowded server with dog crap performance and want to hide it from you so they can give you idiotic excuses; after all, the average user doesn't know shit about how server works, it's so easy to lie to them. My opinion has been, is, and will be to stay away from them.

I can and will update the connectors with an option to disable system information collection, but the important thing to keep in mind is that **the problem lies in the way the server is configured, not the information collection**.

**Comment by drdingens** [comment]
Hey
Thanks again. No I haven’t installed the Hetrix stats agent. 
Maybe I should. I am happy with the Hoster in general, but I see your points. 
Next question:
How to find a good hoster ;-). 
Cheers
Rolf

**Reply by nikosdion** [reply]
My go-to is Rochen. I've had nothing but excellent experience with them for the past 15 years. Their server configuration is top-notch, they are proactive in server management, and when weird bollocks happens they are happy to help.

**Comment by drdingens** [comment]
Thanks. 
Will check them out
Rolf
